ruby_hack_assembler.rb

```diff
@@ -2,4 +2,5 @@
 
 require_relative 'assembler'
 
-puts Assembler.new(ARGF).assemble
```

Comment:
Sorry, my fault! It looks like `ARGF#rewind`, although implemented, doesn’t work properly. TIL.

You could instead read `ARGF` into a `StringIO` and pass that to the assembler, but I understand it’s probably not worth the hassle given that your `File.open(ARGV[0])` solution works. (What’s the `#to_s` for?)

Reply:
OK, I understand what’s happening here. The [documentation](https://www.rubydoc.info/stdlib/core/toplevel#ARGF-constant) says:

>After a file in `ARGV` has been read `ARGF` removes it from the Array. Thus, after all files have been read `ARGV` will be empty.

So once we read to the end of the first (and only) file in `ARGV` by doing `ARGF.readlines`, its name gets _removed_ from `ARGV` and there’s no way to get it back. We can still try to `ARGF.rewind`, but it “[positions the _current file_ to the beginning of input](https://rubydoc.info/stdlib/core/ARGF:rewind)”, so if we’ve already moved past the file argument, there’s either no longer a current file (which is what the CI error means) or we’re now reading from STDIN. Neither of those is what we want.

Sorry for the distraction! I’ve learned a valuable lesson about rewinding `ARGF` today 😂